x-pack/plugins/apm/server/lib/errors/get_error_rate.ts

```diff
@@ -0,0 +1,115 @@
+import {
+  ERROR_GROUP_ID,
+  SERVICE_NAME,
+  PROCESSOR_EVENT,
+} from '../../../common/elasticsearch_fieldnames';
+import { getBucketSize } from '../helpers/get_bucket_size';
+import { rangeFilter } from '../helpers/range_filter';
+import {
+  Setup,
+  SetupTimeRange,
+  SetupUIFilters,
+} from '../helpers/setup_request';
+
+export async function getErrorRate({
+  serviceName,
+  groupId,
+  setup,
+}: {
+  serviceName: string;
+  groupId?: string;
+  setup: Setup & SetupTimeRange & SetupUIFilters;
+}) {
+  const { start, end, uiFiltersES, client, indices } = setup;
+  const { intervalString } = getBucketSize(start, end, 'auto');
+  const groupIdTerm = groupId ? [{ term: { [ERROR_GROUP_ID]: groupId } }] : [];
+
+  const filter = [
+    { term: { [SERVICE_NAME]: serviceName } },
+    { range: rangeFilter(start, end) },
+    ...uiFiltersES,
+  ];
+
+  const aggs = {
+    response_times: {
+      date_histogram: {
+        field: '@timestamp',
+        fixed_interval: intervalString,
+        min_doc_count: 0,
+        extended_bounds: { min: start, max: end },
+      },
+    },
+  };
+
+  const getTransactionBucketAggregation = async () => {
+    const resp = await client.search({
+      index: indices['apm_oss.transactionIndices'],
+      body: {
+        size: 0,
+        query: {
+          bool: {
+            filter: [
+              ...filter,
+              { term: { [PROCESSOR_EVENT]: ProcessorEvent.transaction } },
+            ],
+          },
+        },
+        aggs,
+      },
+    });
+    return {
+      totalHits: resp.hits.total.value,
+      aggs: resp.aggregations?.response_times.buckets,
+    };
+  };
+  const getErrorBucketAggregation = async () => {
+    const resp = await client.search({
+      index: indices['apm_oss.errorIndices'],
+      body: {
+        size: 0,
+        query: {
+          bool: {
+            filter: [
+              ...filter,
+              ...groupIdTerm,
+              { term: { [PROCESSOR_EVENT]: ProcessorEvent.error } },
+            ],
+          },
+        },
+        aggs,
+      },
+    });
+    return resp.aggregations?.response_times.buckets;
+  };
+
+  // Needed to the wrap the call to client.search in a function to avoid "Type instantiation is excessively deep and possibly infinite"
+  const [transactions, errors] = await Promise.all([
+    getTransactionBucketAggregation(),
+    getErrorBucketAggregation(),
+  ]);
+
+  const transactionByTimestamp: Record<number, number> = {};
+  if (transactions?.aggs) {
+    transactions.aggs.forEach((bucket) => {
+      transactionByTimestamp[bucket.key] = bucket.doc_count;
+    });
+  }
+
+  return errors?.map((bucket) => {
+    const { key, doc_count: errorCount } = bucket;
+    return {
+      x: key,
+      y:
+        // If the transaction count doesn't have any hits, it means that the there's no result
+        transactions?.totalHits === 0
+          ? undefined
+          : errorCount / (transactionByTimestamp[key] || 1),
```

Comment:
I have a little trouble considering all the possible cases here. Maybe we can break it up a bit? e.g.:

```ts

const transactionCount = transactionByTimestamp[key] || 0;
const relativeRate = Math.min(errorCount/transactionCount, 0);

return {
  x: key,
  y: relativeRate
};
```

Reply:
I will keep this condition though: 
```js
y: transactions?.totalHits === 0 ? undefined : relativeRate,
```

In case there are no transaction hits I should not return the `y` property. It'll make the Empty state massage to show up on the Error rate chart.